core/ipc/twin/common/src/main/java/org/opennms/core/ipc/twin/common/AbstractTwinSubscriber.java

```diff
@@ -230,72 +136,183 @@ protected TwinUpdate mapTwinResponseToProto(byte[] responseBytes) {
+
+            // Call all consumers
+            this.consumers.forEach(c -> c.accept(this.value.value));
         }
 
-        @Override
-        public int hashCode() {
-            return Objects.hash(key, consumer, clazz);
+        private synchronized void request() {
+            // Send a request
+            final var request = new TwinRequest(this.key, AbstractTwinSubscriber.this.identity.getLocation());
+            AbstractTwinSubscriber.this.sendRpcRequest(request);
+
+            // Cancel an outstanding retry
+            if (this.retry != null) {
+                this.retry.cancel(false);
+            }
+
+            // Schedule a retry
+            this.retry = AbstractTwinSubscriber.this.executorService.schedule(this::request, 5, TimeUnit.SECONDS);
         }
 
-        @Override
-        public String toString() {
-            return new StringJoiner(", ", SessionImpl.class.getSimpleName() + "[", "]")
-                    .add("key='" + key + "'")
-                    .add("consumer=" + consumer)
-                    .add("clazz=" + clazz)
-                    .toString();
+        public synchronized void update(final TwinUpdate update) throws IOException {
+            if (this.value == null || !Objects.equals(this.value.sessionId, update.getSessionId())) {
+                // Either there was no previous known value or the session has restarted
+
+                if (!update.isPatch()) {
+                    this.accept(new Value(update.getSessionId(),
+                                          update.getVersion(),
+                                          AbstractTwinSubscriber.this.objectMapper.readTree(update.getObject())));
+                } else {
+                    this.request();
+                }
+
+            } else {
+                // Same session
+
+                // Ignore update if version is not advancing
+                if (update.getVersion() <= this.value.version) {
+                    return;
+                }
+
+                if (!update.isPatch()) {
+                    this.accept(new Value(update.getSessionId(),
+                                          update.getVersion(),
+                                          AbstractTwinSubscriber.this.objectMapper.readTree(update.getObject())));
+                } else {
+                    if (update.getVersion() == this.value.version + 1) {
+                        // Version advanced - apply path
+                        try {
+                            final var patchObj = AbstractTwinSubscriber.this.objectMapper.readTree(update.getObject());
+                            final var patch = JsonPatch.fromJson(patchObj);
+
+                            final var value = patch.apply(this.value.value);
+
+                            this.accept(new Value(update.getSessionId(), update.getVersion(), value));
+                        } catch (JsonPatchException e) {
+                            throw new IOException("Unable to apply patch", e);
```

Comment:
Instead of throwing an `IOException` you could log the problem and just call `this.request()` like in the other cases where the value could not be processed.

Reply:
This is clearly an error case related to an internal protocol error - not something we would expect on a regular basis. Therefore handling it the same as other decoding errors.

Reply:
I see no benefit in splitting the processing logic of updates into two places. However, this is a matter of taste.